once/__init__.py

```diff
@@ -24,30 +26,57 @@ class _OnceBase(abc.ABC):
     """Abstract Base Class for once function decorators."""
 
     def __init__(self, func: collections.abc.Callable):
-        self._inspect_function(func)
         functools.update_wrapper(self, func)
-        self.lock = _new_lock()
+        self.func = self._inspect_function(func)
         self.called = False
         self.return_value: typing.Any = None
-        self.func = func
+        self.is_asyncgen = inspect.isasyncgenfunction(self.func)
+        self.is_async = True if self.is_asyncgen else inspect.iscoroutinefunction(self.func)
```

Comment:
Maybe a comment about why you chose iscoroutinefunction vs something like isawaitable would be good